### DISCUSSION TITLE
정규화에 대한 설명과 장단점?

### DISCUSSION THREAD

**Discussion by J-I-H-O** [2024-woowacourse-study/level-interview#188] [Database]
.

**Comment by nayonsoso** [comment]
### 정규화의 목적
- 데이터 중복 최소화
- 이상현상 방지
    - 갱신 이상 : 중복된 데이터 중 일부만 갱신하여 데이터의 불일치 발생
    - 삽입 이상 : 불필요한 정보를 함께 저장하지 않고서는 어떤 데이터를 저장하는 것이 불가능
    - 삭제 이상 : 필요한 정보를 함께 삭제하지 않고서는 어떤 정보를 삭제하는 것이 불가능
- 다양한 쿼리를 지원해서 효과적인 검색을 하기 위해
- 데이터 무결성 유지 (정규화로 테이블을 분리하여 제약 조건을 걸면, 무결성을 유지할 수 있다)

### 정규화의 장점
- 데이터 무결성을 해치는 이상 현상을 없앨 수 있다.
- 데이터의 중복을 줄여 저장 공간을 효율적으로 사용할 수 있다.

### 정규화의 단점
- JOIN 연산 증가로 읽기 시간이 길어진다.
- 조회가 핵심 기능을 할 때는 읽기 시간 최적화를 위해 `반정규화 테이블`을 쓸 수 있다.

**Comment by gyungchan-jo** [comment]
정규화는 데이터베이스 설계에서 데이터의 중복을 최소화하고, 데이터 무결성을 유지하기 위해 데이터베이스 테이블을 구조화하는 과정입니다. 정규화는 여러 단계로 나뉘며, 각 단계는 특정 규칙에 따라 테이블을 재구조화합니다. 

정규화의 종류에는 1NF, 2NF, 3NF, BCNF(보이스-코드 정규화), 4NF, 5NF가 있지만, 비공식적으로 3NF면 정규화되었다고 한다.

1. 제1정규형 (1NF)
- 규칙: 모든 속성 값이 원자 값을 가져야 합니다.

2. 제2정규형 (2NF)
- 규칙: 1NF를 만족하고, 부분함수 종속을 제거해야 한다.

3. 제3정규형 (3NF)
- 규칙: 2NF를 만족하고, 이행함수종속을 제거해야 한다.

정규화의 장점
- 데이터 중복 감소
- 데이터 무결성 유지
- 데이터베이스 유지보수 용이
- 효율적 데이터 검색

정규화의 단점
- 조인 연산으로 인한 성능 저하
- 복잡한 쿼리
- 설계 및 관리의 어려움

[참고 블로그](https://wouldyou.tistory.com/111)
